### PR TITLE
feat(framework) Add get `TaskIns` from `run_id` to `LinkState`

### PR DESCRIPTION
**Pull request by chongshenng**
Required for deleting `TaskIns`/`TaskRes` from `LinkState` for a stopped `run_id` in `ExecServicer`.

Merge before:
* #4684

### CODE REVIEW

File: src/py/flwr/server/superlink/linkstate/linkstate_test.py [hunk 1]
```diff
@@ -353,6 +353,43 @@ def test_store_and_delete_tasks(self) -> None:
...
+        assert task_id_2
+
+        expected_task_ids = {task_id_0, task_id_1}
+
+        # Execute
+        result = state.get_task_ids_from_run_id(run_id_0)
+
+        assert result
+        assert result == expected_task_ids
+        self.assertNotIn(task_id_2, result)
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/linkstate_test.py:391]:
```suggestion
        self.assertSetEqual(result, expected_task_ids)
```

Comment (jafermarq) [reply src/py/flwr/server/superlink/linkstate/linkstate_test.py:391]: we could use the built-in mechanism to check two sets being equal

File: src/py/flwr/server/superlink/linkstate/linkstate_test.py [hunk 2]
```diff
@@ -353,6 +353,43 @@ def test_store_and_delete_tasks(self) -> None:
...
+        )
+        task_ins_1 = create_task_ins(
+            consumer_node_id=node_id, anonymous=False, run_id=run_id_0
+        )
+        # Insert a task with a different run_id to ensure it does not appear in result
+        run_id_1 = state.create_run(None, None, "9f86d08", {}, ConfigsRecord())
+        task_ins_2 = create_task_ins(
+            consumer_node_id=node_id, anonymous=False, run_id=run_id_1
+        )
+
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/linkstate_test.py:374]: Should we include a third run that doesn't have any `TaskIns` in it and verify that `get_task_ids_from_run_id` is working as expected?

Comment (chongshenng) [reply src/py/flwr/server/superlink/linkstate/linkstate_test.py:374]: Good suggestion. Added one test for a non-existing `run_id`.

File: src/py/flwr/server/superlink/linkstate/linkstate_test.py [hunk 3]
```diff
@@ -353,6 +353,44 @@ def test_store_and_delete_tasks(self) -> None:
...
+        assert task_id_2
+
+        expected_task_ids = {task_id_0, task_id_1}
+
+        # Execute
+        result = state.get_task_ids_from_run_id(run_id_0)
+        bad_result = state.get_task_ids_from_run_id(15)
+
+        assert result
+        self.assertNotIn(task_id_2, bad_result)
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/linkstate_test.py:391]:
```suggestion
        self.assertEqual(len(bad_result), 0)
```

File: src/py/flwr/server/superlink/linkstate/sqlite_linkstate.py [hunk 4]
```diff
@@ -629,6 +629,25 @@ def delete_tasks(self, task_ids: set[UUID]) -> None:
 
         return None
 
+    def get_task_ids_from_run_id(self, run_id: int) -> set[UUID]:
+        """Return a set of UUID task_ids associated with the given run_id."""
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/sqlite_linkstate.py:633]:
```suggestion
        """Get all TaskIns IDs for the given run_id."""
```

File: src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py [hunk 5]
```diff
@@ -293,6 +293,16 @@ def delete_tasks(self, task_ids: set[UUID]) -> None:
             for task_id in task_res_to_be_deleted:
                 del self.task_res_store[task_id]
 
+    def get_task_ids_from_run_id(self, run_id: int) -> set[UUID]:
+        """Get list of TaskIns IDs for a given run_id."""
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/linkstate/in_memory_linkstate.py:297]:
```suggestion
        """Get all TaskIns IDs for the given run_id."""
```

**Review by jafermarq — APPROVED** [review]
🔥

**Review by danieljanes — APPROVED** [review]

**Pull request merged**